File: container-creation/intern-phxdc-pve1/packer/api/proxmox_upload.py
#!/usr/bin/env python3
import sys, argparse, warnings
from proxmox_utils import get_nodes, get_storages, upload_template, choose_default_storage

# Suppress insecure request warnings (e.g., if PROXMOX_API_URL has verify=False)
warnings.filterwarnings("ignore", category=requests.urllib3.exceptions.InsecureRequestWarning)

def main():
    parser = argparse.ArgumentParser(description="Upload Proxmox LXC template to all nodes.")
    parser.add_argument("--file", required=True, help="Path to the .tar.xz template file.")
    args = parser.parse_args()

    print(f"Starting template upload for: {args.file}")

    try:
        nodes = get_nodes()
        if not nodes:
            print("Error: No Proxmox nodes found.")
            sys.exit(1)

        print(f"Found nodes: {', '.join(nodes)}")

        for node in nodes:
            print(f"--- Processing Node: {node} ---")
            storages_list = get_storages(node)
            
            # Use the utility function to pick the best 'local' storage
            storage = choose_default_storage(storages_list)
            
            if not storage:
                print(f"Warning: No suitable storage found on node {node}. Skipping.")
                continue

            print(f"Uploading to {node}:{storage}...")
            try:
                result = upload_template(node, storage, args.file)
                print(f"Successfully uploaded to {node}:{storage}. Task: {result.get('data')}")
            except Exception as e:
                print(f"Error uploading to {node}:{storage}: {e}")

    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        sys.exit(1)

    print("Template upload process finished.")

if __name__ == "__main__":
    main()

File: packer/api/proxmox_upload.js
#!/usr/bin/env node
// api/proxmox_upload.js
const path = require('path');
const fs = require('fs');
const { getNodes, getStorages, uploadTemplate, chooseDefaultStorage } = require('./proxmox_utils');

function parseArgs() {
  const argv = process.argv.slice(2);
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--file' && argv[i + 1]) {
      out.file = argv[i + 1];
      i++;
    }
  }
  return out;
}

async function main() {
  const args = parseArgs();
  if (!args.file) {
    console.error('Error: --file is required');
    process.exit(1);
  }

  const filepath = args.file;
  if (!fs.existsSync(filepath)) {
    console.error(`Error: file not found: ${filepath}`);
    process.exit(1);
  }

  console.log(`Starting template upload for: ${filepath}`);

  try {
    const nodes = await getNodes();
    if (!nodes || !nodes.length) {
      console.error('Error: No Proxmox nodes found.');
      process.exit(1);
    }

    console.log(`Found nodes: ${nodes.join(', ')}`);

    for (const node of nodes) {
      console.log(`--- Processing Node: ${node} ---`);
      const storagesList = await getStorages(node);
      const storage = chooseDefaultStorage(storagesList);
      if (!storage) {
        console.warn(`Warning: No suitable storage found on node ${node}. Skipping.`);
        continue;
      }

      console.log(`Uploading to ${node}:${storage}...`);
      try {
        const result = await uploadTemplate(node, storage, filepath);
        console.log(`Successfully uploaded to ${node}:${storage}. Task: ${JSON.stringify(result.data || result)}`);
      } catch (e) {
        console.error(`Error uploading to ${node}:${storage}: ${e.message || e}`);
      }
    }
  } catch (e) {
    console.error(`An unexpected error occurred: ${e.message || e}`);
    process.exit(1);
  }

  console.log('Template upload process finished.');
}

if (require.main === module) {
  main();
}
